firebase_admin/app_check.py

```diff
@@ -0,0 +1,146 @@
+# Copyright 2022 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Firebase App Check module."""
+
+from typing import Any, Dict
+import jwt
+from jwt import PyJWKClient, ExpiredSignatureError, InvalidTokenError
+from jwt import InvalidAudienceError, InvalidIssuerError, InvalidSignatureError
+from firebase_admin import _utils
+
+_APP_CHECK_ATTRIBUTE = '_app_check'
+
+def _get_app_check_service(app) -> Any:
+    return _utils.get_app_service(app, _APP_CHECK_ATTRIBUTE, _AppCheckService)
+
+def verify_token(token: str, app=None) -> Dict[str, Any]:
+    """Verifies a Firebase App Check token.
+
+    Args:
+        token: A token from App Check.
+        app: An App instance (optional).
+
+    Returns:
+        Dict[str, Any]: A token's decoded claims if the App Check token
+
+    Raises:
+        ValueError: If ``project_id``, headers, or the decoded token payload is invalid.
+    """
+    return _get_app_check_service(app).verify_token(token)
+
+class _AppCheckService:
+    """Service class that implements Firebase App Check functionality."""
+
+    _APP_CHECK_ISSUER = 'https://firebaseappcheck.googleapis.com/'
+    _JWKS_URL = 'https://firebaseappcheck.googleapis.com/v1/jwks'
+    _project_id = None
+    _scoped_project_id = None
+
+    def __init__(self, app):
+        # Validate and store the project_id to validate the JWT claims
+        self._project_id = app.project_id
+        if not self._project_id:
+            raise ValueError(
+                'Project ID is required to access App Check service. Either set the '
+                'projectId option, or use service account credentials. Alternatively, set the '
+                'GOOGLE_CLOUD_PROJECT environment variable.')
+        self._scoped_project_id = 'projects/' + app.project_id
+
+
+    def verify_token(self, token: str) -> Dict[str, Any]:
+        """Verifies a Firebase App Check token."""
+        _Validators.check_string("app check token", token)
+
+        # Obtain the Firebase App Check Public Keys
+        # Note: It is not recommended to hard code these keys as they rotate,
+        # but you should cache them for up to 6 hours.
+        # Default lifespan is 300 seconds (5 minutes) so we change it to 21600 seconds (6 hours).
+        jwks_client = PyJWKClient(self._JWKS_URL, lifespan=21600)
```

Comment:
Sorry for not catching earlier, I think we should initialize this in the constructor and make jwks_client a class member. Otherwise, every time we call `verify_token()` it will initialize a new instance and I don't think that is necessary plus it will also break the caching mechanism.